core/src/main/scala/kafka/controller/KafkaController.scala

```diff
@@ -284,12 +289,18 @@ class KafkaController(val config : KafkaConfig, zkUtils: ZkUtils, val brokerStat
                   }
                   // If the broker is a follower, updates the isr in ZK and notifies the current leader
                   replicaStateMachine.handleStateChanges(Set(PartitionAndReplica(topicAndPartition.topic,
-                    topicAndPartition.partition, id)), OfflineReplica)
+                    topicAndPartition.partition, id)), OfflineReplica, newBatch = false)
                 }
               }
             }
           }
       }
+
+      inLock(controllerContext.controllerLock) {
```

Comment:
@lindong28 : Hmm, I am not quite sure. Once the controller is done putting the LeaderAndIsr requests in the ControllerChannelManager's in memory queue, it can send a response to the shutting down broker. The response can be sent as long as the network thread is not blocked. So even if there are many pending requests in the request queue of the controller, it typically shouldn't delay the network thread sending a response. Similarly, on the shutting down broker side, the ShutdownRequest is sent in a separate socket channel in a separate thread. The receiving of the ShutdownResponse doesn't go through the request handler threads and should be received immediately even if there are pending requests in the shutting down broker's request queue. So, step (a) can be the bottleneck only if the controller's network threads are blocked for a long time. Is that the case? Typically, this means that the request queue is not configured large enough to match the number of socket connections.